Split tensor

divyanshj16 opened the pull request:
This fixes #383 

#### What does your implementation fix? 
This add split functionality to default tensor type.

#### Explain your changes.
In syft.math it add a split function which uses numpy.array_split() for splitting the array into given axis into `split size` size of each chunk. The function returns a list containing divided sub-arrays. I had to some initial calculations to match the numpy api.

bharathgs commented on syft/tensor.py:
```diff
@@ -3601,3 +3601,33 @@ def zero_(self):
+
+        Parameters
+        ----------
+        tensor: TensorBase
```
change this inline doc (`tensor`?)

bharathgs commented:
looks like few test are missing.

divyanshj16 commented:
Earlier I confused `split_size` variable to be the number of splits. But In the new commit, I changed this!
Actually, split size is chunk size.
If there is an array [0,1,2,3,4,5,6,7,8,9] then if `split_size` is 3 then the array will get divided into subarrays of size 3 and last one with size one as [[0,1,2],[3,4,5],[6,7,8],[9]].
In my earlier implementation the above array would have divided into 3 subarrays [[0,1,2,3],[4,5,6,7],[8,9]]. This has been resolved by this last commit.

bharathgs commented:
your code does not produce the results that you mention in the comment above.
also, please try out the code before a PR.

divyanshj16 commented:
I'll do the changes! I did not understand your second point, `2. in place version of this function is not necessary, however, a split method is(in tensor module)`

bharathgs commented:
change `split_` -> `split`

divyanshj16 commented:
There is already a `split` method in `tensor.py` at line 3041 and `SplitTests` are also there in `test_tensor.py` at line 977. Its tests are not exhaustive. Also, they are not implemented as PyTorch's implementation. 
This was the reason I had to name my test class as `SplitTests2` and my function as `split_`. 
My functions are working good in accordance with the PyTorch's implementation. See the image below.

![image](https://user-images.githubusercontent.com/16683472/32266444-76c3dc2e-bf0d-11e7-92f1-2dd50b7fd32f.png)

I'll add tests in test_math for the split function and will create a PR!
Should I  remove the earlier implementation(i.e implementation on the specified lines)?

bharathgs commented:
ahh, now I see, I missed that. yes remove the earlier implementation

divyanshj16 commented:
Naming conventions are not specific as `axis`(numpy's) and `dim`(pytorch's) are interchangeably used in the whole code. What should I use?

divyanshj16 commented:
I have used `axis` for naming the axis of the tensor. However, we should raise an issue to make it consistent.
